bigtable/google/cloud/bigtable/table.py

```diff
@@ -340,17 +344,19 @@ def yield_rows(self, start_key=None, end_key=None, limit=None,
                         specified row(s). If unset, reads every column in
                         each row.
 
+        :type retry: bool
+        :param retry: (Optional) Apply retry on read rows.
+
         :rtype: :class:`.PartialRowData`
         :returns: A :class:`.PartialRowData` for each row returned
         """
-        request_pb = _create_row_request(
-            self.name, start_key=start_key, end_key=end_key, filter_=filter_,
-            limit=limit)
-        client = self._instance._client
-        response_iterator = client._data_stub.ReadRows(request_pb)
-        # We expect an iterator of `data_messages_v2_pb2.ReadRowsResponse`
-        generator = YieldRowsData(response_iterator)
-        for row in generator.read_rows():
+
+        retryable_read_rows = _RetryableReadRows(
+            self._instance._client, self.name, start_key,
+            end_key, filter_, limit, retry)
+
+        for row in retryable_read_rows().read_rows():
+            retryable_read_rows.last_scanned_key = row.row_key
```

Comment:
I think that`YieldRowData` should keep track of `last_scanned_key` instead of it being tracked in _RetryableReadRows.

`response.last_scanned_row_key` is not generally set.  `YieldRowData` should use `self._row.row_key` in the for loop.  `self._last_scanned_row_key = response.last_scanned_row_key` should only be set after the for loop, only if `response.last_scanned_row_key` is not None or empty

Reply:
Agreed.  After testing, it appears that @sduskis is correct, the ```response.last_scanned_row_key``` is always None (hence, no idea what this was supposed to do).

In our code, ```self._last_scanned_row_key``` can be set to ```self._row.row_key``` after the yield in the following code:
```
                if chunk.commit_row:
                    self._save_current_cell()
                    yield self._row
```
After the yield, the client has been provided with the successfully read row, so it is safe to read from beyond this row key. Please confirm this makes sense, then we will refactor.

Reply:
Sounds reasonable. You should do that.

Reply:
Thanks @jonparrott. Please note that the change to YieldRowsData will be done on PR #4899 since there is validation involved and the changes are localized to that code.  Based on the above comments, we are covered in this PR with the current implementation, where last_scanned_key is maintained in the retry logic.

Reply:
I'm not sure what you mean.

Reply:
Upon re-review with @sduskis, the last_scanned_row_key in a ReadRowsResponse was made clearer:

There are two things in a ReadRowsResponse:
1) a list of chunks
2) a variable literally called last_scanned_row_key

If there are no chunks, then last_scanned_row_key should be set.  See here for the protobuf definition: https://github.com/googleapis/googleapis/blob/master/google/bigtable/v2/bigtable.proto#L172

Based on this, the code is being refactored according to the comment from @sduskis above.